open-trainings/src/main/java/com/smalaca/opentrainings/domain/offeracceptancesaga/OfferAcceptanceSaga.java

```diff
@@ -2,19 +2,46 @@
 
 import com.smalaca.domaindrivendesign.Saga;
 import com.smalaca.opentrainings.application.offer.AcceptOfferCommand;
+import com.smalaca.opentrainings.domain.clock.Clock;
 import com.smalaca.opentrainings.domain.offeracceptancesaga.events.OfferAcceptanceRequestedEvent;
+import com.smalaca.opentrainings.domain.offeracceptancesaga.events.OfferAcceptanceSagaEvent;
 
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.UUID;
+import java.util.function.BiConsumer;
 
 @Saga
 public class OfferAcceptanceSaga {
     private final UUID offerId;
+    private final List<AcceptedEvent> events = new ArrayList<>();
+    private boolean isCompleted;
 
     public OfferAcceptanceSaga(UUID offerId) {
         this.offerId = offerId;
     }
 
-    public AcceptOfferCommand accept(OfferAcceptanceRequestedEvent event) {
+    public AcceptOfferCommand accept(OfferAcceptanceRequestedEvent event, Clock clock) {
+        process(event, clock.now());
         return new AcceptOfferCommand(event.offerId(), event.firstName(), event.lastName(), event.email(), event.discountCode());
     }
+
+    private void process(OfferAcceptanceRequestedEvent event, LocalDateTime consumedAt) {
+        AcceptedEvent acceptedEvent = new AcceptedEvent(event.eventId(), consumedAt, event);
+        events.add(acceptedEvent);
+        isCompleted = true;
+    }
+
+    public boolean isCompleted() {
+        return isCompleted;
+    }
+
+    public void forEachEvent(BiConsumer<OfferAcceptanceSagaEvent, LocalDateTime> consumer) {
```

Comment:
It should be for last event. Those are immutable so it's pointless to save anything but the last